lib/stronger_parameters/controller_support/permitted_parameters.rb

```diff
@@ -0,0 +1,169 @@
+                true
+              else
+                v.dup
+              end
+            end
+          end
+        end
+
+        def sugar(value)
+          case value
+          when Array
+            ActionController::Parameters.array(*value.map { |v| sugar(v) }) | ActionController::Parameters.nil
+          when Hash
+            constraints = value.each_with_object({}) do |(key, v), memo|
+              memo[key] = sugar(v)
+            end
+            ActionController::Parameters.map(constraints)
+          else
+            value
+          end
+        end
+      end
+
+      DEFAULT_PERMITTED = PermittedParametersHash.new(
+        controller: ActionController::Parameters.anything,
+        action: ActionController::Parameters.anything,
+        format: ActionController::Parameters.anything,
+        authenticity_token: ActionController::Parameters.string
+      )
+
+      module ClassMethods
+        def self.extended(base)
+          base.class_eval do
+            class_attribute :log_unpermitted_parameters, instance_accessor: false
+          end
+        end
+
+        def log_unpermitted_parameters!
+          self.log_unpermitted_parameters = true
+        end
+
+        def permitted_parameters(action, permitted)
+          if permitted_parameters_list[action] == :anything && permitted != :anything
+            raise ArgumentError, "#{self}/#{action} can not add to :anything"
+          end
+
+          permitted_parameters_list.deep_merge!(action => permitted)
+        end
+
+        def permitted_parameters_list
+          @permitted_parameters_list ||= if superclass.respond_to?(:permitted_parameters_list)
+            superclass.permitted_parameters_list.deep_dup
+          else
+            all = DEFAULT_PERMITTED.dup
+            all[:test_route] = ActionController::Parameters.string if Rails.env.test?
+            { all: all }.with_indifferent_access
+          end
+        end
+
+        def permitted_parameters_for(action)
+          unless for_action = permitted_parameters_list[action]
+            location = instance_method(action).source_location
+            raise KeyError, "Action #{action} for #{self} does not have any permitted parameters (#{location.join(":")})"
+          end
+          return :anything if for_action == :anything
+          permitted_parameters_list[:all].merge(for_action)
+        end
+      end
+
+      private
+
+      def permit_parameters
+        action = params[:action].to_sym
+        permitted = self.class.permitted_parameters_for(action)
+
+        if permitted == :anything
+          Rails.logger.warn("#{params[:controller]}/#{params[:action]} does not filter parameters")
+          return
+        end
+
+        permitted_params = without_invalid_parameter_exceptions { params.permit(permitted) }
+        restricted_keys = flat_keys(params) - flat_keys(permitted_params)
+        filter_params = !self.class.log_unpermitted_parameters
```

Comment:
call it `log_unpermitted` so we can revert the logic ... would be less negation ..